examples/python/xds/server.py

```diff
@@ -52,46 +62,75 @@ def SayHello(self, request: helloworld_pb2.HelloRequest,
-        xds=secure)
-
-    # Add the application servicer to the server.
-    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(hostname), server)
+def _configure_maintenance_server(server: grpc.Server,
+                                  maintenance_port: int) -> None:
+    listen_address = f"{_LISTEN_HOST}:{maintenance_port}"
+    server.add_insecure_port(listen_address)
 
     # Create a health check servicer. We use the non-blocking implementation
     # to avoid thread starvation.
     health_servicer = health.HealthServicer(
         experimental_non_blocking=True,
-        experimental_thread_pool=futures.ThreadPoolExecutor(max_workers=1))
-    health_pb2_grpc.add_HealthServicer_to_server(health_servicer, server)
+        experimental_thread_pool=futures.ThreadPoolExecutor(
+            max_workers=_THREAD_POOL_SIZE))
 
     # Create a tuple of all of the services we want to export via reflection.
     services = tuple(
         service.full_name
         for service in helloworld_pb2.DESCRIPTOR.services_by_name.values()) + (
             reflection.SERVICE_NAME, health.SERVICE_NAME)
 
-    # Add the reflection service to the server.
+    # Mark all services as healthy.
+    health_pb2_grpc.add_HealthServicer_to_server(health_servicer, server)
+    for service in services:
+        health_servicer.set(service, health_pb2.HealthCheckResponse.SERVING)
     reflection.enable_server_reflection(services, server)
-    listen_address = f"[::]:{port}"
-    if not secure:
+
+
+def _configure_greeter_server(server: grpc.Server, port: int, secure_mode: bool,
+                              hostname) -> None:
+    # Add the application servicer to the server.
+    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(hostname), server)
+    listen_address = f"{_LISTEN_HOST}:{port}"
+    if not secure_mode:
         server.add_insecure_port(listen_address)
     else:
-        print("Running with xDS Server credentials")
+        # Use xDS credentials.
+        logger.info("Running with xDS Server credentials")
+
+        # Fall back to insecure credentials.
         server_fallback_creds = grpc.insecure_server_credentials()
         server_creds = grpc.xds_server_credentials(server_fallback_creds)
         server.add_secure_port(listen_address, server_creds)
-    server.start()
 
-    # Mark all services as healthy.
-    overall_server_health = ""
-    for service in services + (overall_server_health,):
-        health_servicer.set(service, health_pb2.HealthCheckResponse.SERVING)
 
-    # Park the main application thread.
-    server.wait_for_termination()
+def serve(port: int, hostname: str, maintenance_port: int,
+          secure_mode: bool) -> None:
+    if port == maintenance_port:
+        # If maintenance port and port are the same, start a single server.
+        server = grpc.server(
+            futures.ThreadPoolExecutor(max_workers=_THREAD_POOL_SIZE))
+        _configure_greeter_server(server, port, secure_mode, hostname)
```

Comment:
This seems a bit awkward. What happens if the port and maintenance port is the same but secure mode is used? Anyway, this is not important for the example, since we will be asking users to use separate ports.

Reply:
Or do not allow the maintenance port to be specified. It can be regular service port + 1

Reply:
> What happens if the port and maintenance port is the same but secure mode is used?

There's a check for this in before the call to `serve`. It's considered an illegal argument. Maybe there's some aspect of this I'm not getting?

> Or do not allow the maintenance port to be specified. It can be regular service port + 1

We could do this, but we'd probably want to change the other languages to align on their CLI arguments. I'm following C++'s pattern here. That's why I had to write my own bool-parsing function. Absl flags differ in their behavior with respect to bools from `argparse`.

Reply:
> We could do this, but we'd probably want to change the other languages to align on their CLI arguments. I'm following C++'s pattern here. That's why I had to write my own bool-parsing function. Absl flags differ in their behavior with respect to bools from `argparse`.

Java follows the pattern https://github.com/grpc/grpc-java/blob/master/examples/example-xds/src/main/java/io/grpc/examples/helloworldxds/XdsHelloWorldServer.java#L65 so you have a precedent and we can then get C++ to follow that :-)

Reply:
Works for me. Done.